File: leet/valid-parentheses.ts
/**
 * https://leetcode.com/problems/valid-parentheses/description/
 */
export function isValid(s: string): boolean {
  const history: string[] = [];

  for (let char of s) {
    if (char === '(' || char === '[' || char === '{') {
      history.push(char);
    } else {
      const last = history.pop();
      if (last === '(' && char === ')') continue;
      else if (last === '[' && char === ']') continue;
      else if (last === '{' && char === '}') continue;
      return false;
    }
  }

  return history.length === 0;
}

File: returned/python/valid_parentheses.py
# https://leetcode.com/problems/valid-parentheses/description/

class Solution:
    def isValid(self, s: str) -> bool:
        progress = []
        for b in s:
            if b == '{' or b == '(' or b == '[':
                progress.append(b)
            elif len(progress) == 0:
                return False
            elif b == '}' and progress.pop() != '{':
                return False
            elif b == ')' and progress.pop() != '(':
                return False
            elif b == ']' and progress.pop() != '[':
                return False
        return len(progress) == 0
